crates/derive/src/stages/batch_queue.rs

```diff
@@ -1 +1,369 @@
+    next_spans: Vec<SingleBatch>,
+
+    /// Used to validate the batches.
+    fetcher: BF,
+}
+
+impl<DAP, CP, BF> BatchQueue<DAP, CP, BF>
+where
+    DAP: DataAvailabilityProvider + Debug,
+    CP: ChainProvider + Debug,
+    BF: SafeBlockFetcher + Debug,
+{
+    /// Creates a new [BatchQueue] stage.
+    pub fn new(cfg: RollupConfig, prev: ChannelReader<DAP, CP>, fetcher: BF) -> Self {
+        Self {
+            cfg,
+            prev,
+            origin: None,
+            l1_blocks: Vec::new(),
+            batches: Vec::new(),
+            next_spans: Vec::new(),
+            fetcher,
+        }
+    }
+
+    /// Returns the L1 origin [BlockInfo].
+    pub fn origin(&self) -> Option<&BlockInfo> {
+        self.prev.origin()
+    }
+
+    /// Pops the next batch from the current queued up span-batch cache.
+    /// The parent is used to set the parent hash of the batch.
+    /// The parent is verified when the batch is later validated.
+    pub fn pop_next_batch(&mut self, parent: L2BlockRef) -> Option<SingleBatch> {
+        if self.next_spans.is_empty() {
+            return None;
+        }
+        let mut next = self.next_spans.remove(0);
+        next.parent_hash = parent.info.hash;
+        Some(next)
+    }
+
+    /// Returns the next valid batch upon the given safe head.
+    /// Also returns the boolean that indicates if the batch is the last block in the batch.
+    pub async fn next_batch(&mut self, parent: L2BlockRef) -> StageResult<SingleBatch> {
+        if !self.next_spans.is_empty() {
+            // There are cached singular batches derived from the span batch.
+            // Check if the next cached batch matches the given parent block.
+            if self.next_spans[0].timestamp == parent.info.timestamp + self.cfg.block_time {
+                return self
+                    .pop_next_batch(parent)
+                    .ok_or(anyhow!("failed to pop next batch from span batch").into());
+            }
+            // Parent block does not match the next batch.
+            // Means the previously returned batch is invalid.
+            // Drop cached batches and find another batch.
+            self.next_spans.clear();
+            // TODO: log that the provided parent block does not match the next batch.
+            // TODO: metrice the internal batch drop.
+        }
+
+        // If the epoch is advanced, update the l1 blocks.
+        // Advancing epoch must be done after the pipeline successfully applies the entire span
+        // batch to the chain.
+        // Because the span batch can be reverted during processing the batch, then we must
+        // preserve existing l1 blocks to verify the epochs of the next candidate batch.
+        if !self.l1_blocks.is_empty() && parent.l1_origin.number > self.l1_blocks[0].number {
+            for (i, block) in self.l1_blocks.iter().enumerate() {
+                if parent.l1_origin.number == block.number {
+                    self.l1_blocks.drain(0..=i);
```

Comment:
Think we'd like to keep `i` here
```suggestion
                    self.l1_blocks.drain(0..i);
```

Reply:
Good catch!